### PR TITLE
[Wallet][ReDesign] Name and number screen

### PR DESCRIPTION
**Pull request by jeanregisser**
### Description

This implements the [name and number screen redesign](https://www.figma.com/file/zt7aTQ5wuXycIwxq5oAmF9/Wallet--Refresh?node-id=5404%3A399).

Main highlights:
- `<PhoneNumberInput />` has been refactored and is hopefully now much simpler to reason about. It's been turned into a controlled component.
- Default country is set using the device locale setting.
- Number placeholder changes according to the selected country
- Language selection is now done in a dedicated screen (displayed with the new modal card style on iOS).
- Countries can now also be searched using their 2 or 3 letters ISO code. This makes it possible to find `United States` by just entering `us`.
- Similar to the previous implementation on Android, we ask the user for permission to read the device phone number when focusing the phone number input or pressing the country button.
- Performance should be hopefully improved when typing a number quickly (no more redux dispatch and iterating through the countries when typing).

### Other changes

- Fix position of `<DevSkipButton />` on iOS devices with notch

### Tested

name-and-number

### Related issues

- Epic #3645

### Backwards compatibility

Yes

### CODE REVIEW

File: packages/mobile/src/onboarding/registration/JoinCelo.tsx [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+  state: State = {
+    name: this.props.cachedName || '',
+    ...getPhoneNumberState(
+      this.props.cachedNumber || '',
+      this.props.cachedCountryCallingCode || '',
+      RNLocalize.getCountry()
+    ),
+  }
+
+  UNSAFE_componentWillReceiveProps(nextProps: Props) {
```
Comment (jeanregisser) [inline packages/mobile/src/onboarding/registration/JoinCelo.tsx]: Don't judge me 😅 
This will be eliminated by switching to a function component. 
I wanted to keep the class component here for now to prevent unnecessary changes from the existing code.

File: packages/mobile/src/navigator/Headers.v2.tsx [hunk 2]
```diff
@@ -97,6 +100,12 @@ export const headerWithCancelButton: StackNavigationOptions = {
   headerLeft: () => <CancelButton />,
 }
 
+export const headerWithCloseButton: StackNavigationOptions = {
+  ...emptyHeader,
+  headerLeft: () => <TopBarIconButton icon={<Times />} onPress={navigateBack} />,
+  headerLeftContainerStyle: { paddingLeft: 20 },
```
Comment (aaronmgdr) [inline packages/mobile/src/navigator/Headers.v2.tsx:111]: my understand is that paddings are 8, 12, 16 or 24. 

also how about having this style be defined in a Stylesheet.create?

File: packages/mobile/src/onboarding/registration/SelectCountry.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  const onSelect = useCallback(
+    (country: LocalizedCountry) => {
+      navigation.navigate(Screens.JoinCelo, { selectedCountryCodeAlpha2: country.alpha2 })
+
+      // Wait for next frame before navigating back
+      // so the user can see the changed selection briefly
+      // requestAnimationFrame(() => {
+      //   navigateBack()
```
Comment (aaronmgdr) [inline packages/mobile/src/onboarding/registration/SelectCountry.tsx]: ??

Comment (jeanregisser) [reply packages/mobile/src/onboarding/registration/SelectCountry.tsx]: Oops, obsolete comment, removed 👍

File: packages/mobile/src/navigator/Headers.v2.tsx (hunk 2 above)
Comment (jeanregisser) [reply packages/mobile/src/navigator/Headers.v2.tsx:111]: Yes it should be one of these sizes, however it felt a bit too far on the right in the iOS modal style.
We can revisit later.

File: packages/mobile/src/onboarding/registration/JoinCelo.tsx (hunk 1 above)
Comment (jmrossy) [reply packages/mobile/src/onboarding/registration/JoinCelo.tsx]: Why not put it in componentDidUpdate? Though we waste a render, we avoid using a deprecated method

**Review by jmrossy — APPROVED** [review]
Looks great to me!

File: packages/mobile/src/onboarding/registration/JoinCelo.tsx (hunk 1 above)
Comment (jeanregisser) [reply packages/mobile/src/onboarding/registration/JoinCelo.tsx]: Good point. Changing this.

File: packages/mobile/src/onboarding/registration/JoinCelo.tsx [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    const countryCallingCode = this.countries.getCountryByCodeAlpha2(countryCodeAlpha2)
+      ?.countryCallingCode
+    const { cachedName, cachedNumber, cachedCountryCallingCode } = this.props
+
+    if (
+      cachedName === name &&
+      cachedNumber === e164Number &&
+      cachedCountryCallingCode === countryCallingCode
+    ) {
+      this.goToNextScreen()
```
Comment (i1skn) [inline packages/mobile/src/onboarding/registration/JoinCelo.tsx:165]: Should we add `this.props.setPromptForno(true)` here as well or it's cached in redux-persist?

File: packages/mobile/src/onboarding/registration/JoinCelo.tsx [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  container: {
+    flex: 1,
+    justifyContent: 'space-between',
+    backgroundColor: colors.onboardingBackground,
+  },
+  scrollContainer: {
+    padding: 24,
+    paddingTop: 80,
+  },
+  nameInput: {},
```
Comment (i1skn) [inline packages/mobile/src/onboarding/registration/JoinCelo.tsx]: I guess, we can remove this 🙂

File: packages/mobile/src/onboarding/registration/SelectCountry.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+  const inset = useSafeArea()
+
+  return (
+    <View style={styles.container}>
+      <View style={styles.searchInputContainer}>
+        <SearchInput
+          placeholder={t('global:search')}
+          value={searchText}
+          onChangeText={onChangeText}
```
Comment (i1skn) [inline packages/mobile/src/onboarding/registration/SelectCountry.tsx]: Can we use here `setSearchText` directly without `onChangeText` function?

File: packages/mobile/src/onboarding/registration/SelectCountry.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  function onChangeText(text: string) {
+    setSearchText(text)
+  }
+
+  const onSelect = useCallback(
+    (country: LocalizedCountry) => {
+      navigation.navigate(Screens.JoinCelo, { selectedCountryCodeAlpha2: country.alpha2 })
+    },
+    [dispatch]
+  )
```
Comment (i1skn) [inline packages/mobile/src/onboarding/registration/SelectCountry.tsx]: I think here we can do:
```suggestion
function onSelect(country: LocalizedCountry) {
   navigation.navigate(Screens.JoinCelo, { selectedCountryCodeAlpha2: country.alpha2 })
 }
```

File: packages/react-components/components/FormField.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  return (
+    <View style={[styles.container, style]}>
+      <FormLabel style={styles.label}>{label}</FormLabel>
+      {children}
+    </View>
+  )
+}
+
+const styles = StyleSheet.create({
+  container: {},
```
Comment (i1skn) [inline packages/react-components/components/FormField.tsx]: I guess, we can remove this

File: packages/mobile/src/onboarding/registration/JoinCelo.tsx (hunk 4 above)
Comment (jeanregisser) [reply packages/mobile/src/onboarding/registration/JoinCelo.tsx:165]: It's persisted.

**Review by i1skn — COMMENTED** [review]
Overall looks very good and thanks for such a tremendous effort!

File: packages/mobile/src/onboarding/registration/JoinCelo.tsx (hunk 5 above)
Comment (jeanregisser) [reply packages/mobile/src/onboarding/registration/JoinCelo.tsx]: Yep, thanks.

File: packages/mobile/src/onboarding/registration/SelectCountry.tsx (hunk 6 above)
Comment (jeanregisser) [reply packages/mobile/src/onboarding/registration/SelectCountry.tsx]: Yes, thanks 👍

File: packages/mobile/src/onboarding/registration/SelectCountry.tsx (hunk 7 above)
Comment (jeanregisser) [reply packages/mobile/src/onboarding/registration/SelectCountry.tsx]: Indeed 👍

File: packages/react-components/components/FormField.tsx (hunk 8 above)
Comment (jeanregisser) [reply packages/react-components/components/FormField.tsx]: 👍
